### PR TITLE
Malmo MDP Support

### PR DESCRIPTION
**Pull request by howard-abrams**
Added some basic Malmo MDP support in a similar pattern to the Gym support already there.

With the exception of the external dependency, should be fairly straight forward, but please let me know if I missed something. 

Also, I have an example ready in dl4j-examples, but I didn't want to create that pull request until this one had landed. You can check it out from my fork if you'd like to try this code out.

### CODE REVIEW

**Comment by arronmabrey** [comment]
I didn't look at the changes, but when I found this lib, I thought about the [malmo](https://github.com/Microsoft/malmo) support! Thanks for putting it together!

File: rl4j-malmo/pom.xml [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+  </properties>
+    
+    <dependencies>
+        <dependency>
+            <groupId>org.deeplearning4j</groupId>
+            <artifactId>rl4j-api</artifactId>
+            <version>${project.version}</version>
+        </dependency>
+        <dependency>
+        	<groupId>com.microsoft.msr</groupId>
```
Comment (agibsonccc) [inline rl4j-malmo/pom.xml]: I can't really justify a system jar in anything I release to maven central. That's a QA disaster waiting to happen. Is there another way we could do this? Maybe publish to maven central?

File: rl4j-malmo/src/main/java/org/deeplearning4j/malmo/MalmoActionSpace.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+ * Abstract base class, based on DiscreteSpace for all Malmo-specific action spaces
+ *
+ */
+public abstract class MalmoActionSpace extends DiscreteSpace
+{
+	protected String[] actions;
+	
+	public MalmoActionSpace(int size)
+	{
+		super(size);
```
Comment (agibsonccc) [inline rl4j-malmo/src/main/java/org/deeplearning4j/malmo/MalmoActionSpace.java]: No c style braces.

File: rl4j-malmo/src/main/java/org/deeplearning4j/malmo/MalmoActionSpaceDiscrete.java [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+ * 
+ * @author howard-abrams (howard.abrams@ca.com) on 1/12/17.
+ *
+ * Action space that allows for a fixed set of specific actions
+ *
+ */
+public class MalmoActionSpaceDiscrete extends MalmoActionSpace
+{
+	public MalmoActionSpaceDiscrete( String... actions )
+	{
```
Comment (agibsonccc) [inline rl4j-malmo/src/main/java/org/deeplearning4j/malmo/MalmoActionSpaceDiscrete.java]: Braces.

File: rl4j-malmo/src/main/java/org/deeplearning4j/malmo/MalmoBox.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+/**
+ * 
+ * @author howard-abrams (howard.abrams@ca.com) on 1/12/17.
+ *
+ * Simple value array similar to Gym Box model, but without a JSON constructor
+ *
+ */
+public class MalmoBox implements Encodable
+{
+	double[] value;
```
Comment (agibsonccc) [inline rl4j-malmo/src/main/java/org/deeplearning4j/malmo/MalmoBox.java]: No c style braces.

File: rl4j-malmo/src/main/java/org/deeplearning4j/malmo/MalmoDescretePositionFramePolicy.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+ *
+ */
+public class MalmoDescretePositionFramePolicy implements MalmoFramePolicy
+{
+	MalmoObservationSpacePosition observationSpace = new MalmoObservationSpacePosition();
+	
+	@Override
+	public boolean isFrameConsistant( WorldState world_state, WorldState original_world_state )
+	{
+        MalmoBox last_observation = observationSpace.getObservation( world_state );
```
Comment (agibsonccc) [inline rl4j-malmo/src/main/java/org/deeplearning4j/malmo/MalmoDescretePositionFramePolicy.java]: Get rid of the spaces here.

File: rl4j-malmo/src/main/java/org/deeplearning4j/malmo/MalmoEnv.java [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+	// Malmo Java client library depends on native library
+    static
+    {
+    	String malmoHome = System.getenv("MALMO_HOME");
+    	
+    	if ( malmoHome == null )
+    		throw new RuntimeException("MALMO_HOME must be set to your Malmo environement.");
+
+    	System.load( malmoHome + "/Java_Examples/libMalmoJava.jnilib");
+//        System.loadLibrary( malmoHome + "/Java_Examples/MalmoJava");
```
Comment (agibsonccc) [inline rl4j-malmo/src/main/java/org/deeplearning4j/malmo/MalmoEnv.java]: Bit concerned here. I would find a better way to do this. I would bundle the native resource *with* the library.

File: rl4j-malmo/src/main/java/org/deeplearning4j/malmo/MalmoEnv.java [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+		}
+
+		agent_host = null;
+	}
+
+	@Override
+	public StepReply<MalmoBox> step( Integer action )
+	{
+		agent_host.sendCommand( (String)actionSpace.encode( action ) );
+
```
Comment (agibsonccc) [inline rl4j-malmo/src/main/java/org/deeplearning4j/malmo/MalmoEnv.java]: Braces.

**Comment by howard-abrams** [comment]
I know what you mean about the system jar and the native library. Unfortunately that's out of my control. I can attempt to talk to the Microsoft/Malmo folks, but java is a very small part of what they provide so it might be an uphill battle talking them into publishing to maven.  Any suggestions?

In terms of braces, spaces, is there a formatter (eclipse or otherwise) that I can use to format the code to match the project? The code formatting that was already in rl4j/examples seemed inconsistent so I just used what I liked.

**Comment by agibsonccc** [comment]
@huitseeker I believe was working on an auto formatter. As for the native library we can work around that by bundling it as a classpath resource under src/main/resources - for the system jar, there's a few things we can do here: Deploy the jar to maven central here: http://www.sonatype.org/nexus/2015/06/02/how-to-publish-software-artifacts-to-maven-central/ *or* bundle the source code of the project with this project and build it like that.

 I mean no offense here - but we submit things to maven central, and in order to do that and not have this be a big headache we need things to hit certain standards.

**Comment by howard-abrams** [comment]
No offense taken - I knew this would be an issue, but thought I'd use the pull request to figure out what the best options were. Let me try a few things and update the code.

Meanwhile, @huitseeker, if you have a formatter that would be great.

**Comment by howard-abrams** [comment]
Thinking more, I'd want to stay away from publishing someone else's stuff if we can help it. I've create an issue for Malmo; we'll see what they say:

https://github.com/Microsoft/malmo/issues/469

**Comment by huitseeker** [comment]
Thanks for the contribution @howard-abrams! The automated syntax formatter is available in the current master, it just wasn't at the time of your PR.

You can use it using `mvn formatter:format` at the root of the repo and it will take into account all the submodules explicitly listed in the configuration of the formatter plugin (see the root pom).

To show how this works I've created a branch with a few commits :
- your rebased, squashed commits,
- a bump of the project's version #
- adding your submodule to the root pom file
- applying automated formatting 

It's stored here:
https://github.com/deeplearning4j/rl4j/compare/master...huitseeker:pr/21?expand=1

**Comment by howard-abrams** [comment]
Thanks! - I'll take a look

**Comment by howard-abrams** [comment]
Ok, I've used the formatter and squashed/rebased.  Still have to get the maven dependency stuff worked out with the Malmo team, but hopefully @agibsonccc 's comment on that issue will remind them about it.

In the meantime, I may add/tweak a few things while we wait.

**Comment by agibsonccc** [comment]
Thanks! We'll continue lobbying there. I'm glad they're receptive at least!

**Comment by howard-abrams** [comment]
Ok, I went ahead and ask the maven central folks to publish the jar dependency. If the malmo folks are interested in publishing future releases that would be great, but in the meantime, I think this is unblocked.

@agibsonccc - Please take a look at the latest diffs and let me know if there is anything you need fixed up before this is ready for merge.

**Comment by saudet** [comment]
@howard-abrams Great! Thanks for that. Do we still need to install it manually as per the instructions you added in the README.md file though?

**Comment by howard-abrams** [comment]
@agibsonccc, yes.  The java build/run dependency is now managed via maven, but to have the sample(s) actual work, you'll still need malmo - it's is basically a complete copy of Minecraft with a special mod.

**Comment by saudet** [comment]
Ok, so basically we need to run the modified game. Looks good to me then!

**Pull request merged**